#### Problem 4:

A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

In [1]:
import numpy as np

In [2]:
# Determine if number is a palindrome
def is_palindrome(number):
    # Test for palindrome by turning number into a string
    # and comparing it to it's reverse.
    # [::-1] = start at the end, end at the start, and move in steps of -1 = reverse
    return str(number) == str(number)[::-1]

In [3]:
# Calculate list of all palindromes
def find_palindromes():
    palindromes = []
    products = []
    for first_multiplier in range(100, 1000):
        for second_multiplier in range(100, 1000):
            product = first_multiplier * second_multiplier
            if is_palindrome(product):
                palindromes.append(product)
                products.append([first_multiplier, second_multiplier])
    return palindromes, products

In [4]:
results, products = find_palindromes()

solution = np.max(results)
solution_products = products[results.index(solution)]

In [5]:
print(f"The largest palindrome made from the product of two 3-digit numbers = {solution_products[0]} x {solution_products[1]} = {solution:,}")

The largest palindrome made from the product of two 3-digit numbers = 913 x 993 = 906,609


In [6]:
# Calculate list of all palindromes that are products of two numbers of specified digits:
def find_palindromes_generic(first_number_digits, second_number_digits):
    palindromes = []
    products = []
    for first_multiplier in range(10**(first_number_digits-1), 10**(first_number_digits)):
        for second_multiplier in range(10**(second_number_digits-1), 10**(second_number_digits)):
            product = first_multiplier * second_multiplier
            if is_palindrome(product):
                palindromes.append(product)
                products.append([first_multiplier, second_multiplier])
    return palindromes, products

In [7]:
first_number_digits = 4
second_number_digits = 4
results, products = find_palindromes_generic(first_number_digits, second_number_digits)

solution = np.max(results)
solution_products = products[results.index(solution)]

print(f"The largest palindrome made from the product of a {first_number_digits}-digit number and a {second_number_digits}-digit number = {solution_products[0]} x {solution_products[1]} = {solution:,}")

The largest palindrome made from the product of a 4-digit number and a 4-digit number = 9901 x 9999 = 99,000,099


Problem 4 Solution: The largest palindrom made from the product of two 3-digit numbers = 913 x 993 = 906,609.